# Call Somatic Structural Variant Pipeline

* [Overview](#Overview)
* [How To Run](#How-To-Run)
* [Flow Diagram](#flow-diagram)
* [Pipeline Steps](#pipeline-steps)
    1. [Calling Structural Variants](#somatic-sv)
* [Inputs](#Inputs)
* [Outputs](#outputs)
* [Testing and Validation](#testing-and-validation)
    * [Test Data Set](#test-data-set)
    * [Performance Validation](#performance-validation)
    * [Validation Tool](#validation-tool)


## Overview:
The call-sSV pipeline calls structural variants utilizing [Delly](https://github.com/dellytools/delly). This pipeline requires at least one tumor sample and a matched control sample.
This pipeline is developed using Nextflow , docker and can run either on a single node linux machine or a multi-node HPC cluster (e.g. Slurm, SGE). Additionally, it has been validated with the SMC-HET dataset.

## How to Run:

1.	Make sure the pipeline is already downloaded to your machine. You can either download the stable release or the dev version by cloning the repo.
2.	Update the nextflow.config file for input, output, and parameters. An example can be found [here](https://github.com/uclahs-cds/pipeline-call-sSV/blob/main/pipeline/config/nextflow.config). See [Inputs](#inputs) for the description of each variables in the config file.
3.	Update the input CSV. See [Inputs](#inputs) for the columns needed. All columns must exist to run the pipeline. An example can be found [here](https://github.com/uclahs-cds/pipeline-call-sSV/blob/main/pipeline/input/paired_turmor_control_samples.csv).
4.	See the submission script, [here](https://github.com/uclahs-cds/tool-submit-nf), to submit your pipeline

## Flow Diagram:

### Regenotype gSV:

![](https://github.com/uclahs-cds/pipeline-regenotype-gSV/blob/dev-documentation-ghouse/Regenotype-gSV-workflow.PNG)

## Pipeline Steps:

#### 1. Calling Single Sample Somatic Structural Variants
The first step of the pipeline requires aligned, sorted tumor sample .bam file and a matched control sample. A single SV .bcf file is generated by calling the Delly Merge command and passing BCF files of all the samples.

#### 2. Calling Structural Variants – Joint Genotyping
The second step of the pipeline genotypes the single unified list generated above for all the input sample BAM files. This can be run in parallel for each sample.To do this, Delly requires an aligned and sorted BAM file and BAM index as an input, as well as the merged BCF output from the step 1.
Currently the following filters are applied:
•	Map-qual: >= 20 (Applied / parameterized)

#### 3. Merge all the genotypes samples generated in Step 2 above into a single VCF/BCF using BCFtools merge.

## Inputs

### Input CSV

The input CSV should have all columns below and in the same order. An example of the input CSV can be found here.

| Field |	Type |	Description |
|--- | --- | --- |
|patient |	String |	The patient name to be passed to the final BCF.No white space allowed. |
|sample	| Varchar	| The sample name to be passed to the final BCF. No white space allowed. |
|input_bam	| Varchar	| Absolute path to the BAM file for the sample. |
|input_sv_bcf |	Varchar	| Absolute path to the sv BCF file (generated by the call sv process). |
|input_cnv_bcf |	Varchar	| Absolute path to the cnv BCF file (generated by the call cnv process). |

## Nextflow Config File Parameters
| Input Parameter |	Required |	Type |	Description |
| ------- |   --------- | ------ | -------------|
| dataset_id |	yes	| String |	Boutros lab dataset id |
| blcds_registered_dataset	| yes |	Boolean | Affirms if dataset should be registered in the Boutros Lab Data registry. Default value is false. |
| sge_scheduler	| Yes	| Boolean	| Affirms whether job will be executed on the SGE cluster. Default value is false. |
| input_bams_bcfs |	Yes |	String	| Absolute path to the input CSV file for the pipeline. |
| reference_fasta	| Yes |	Path	| Absolute path to the reference genome fasta file. The reference genome is used by Delly for structural variant calling. |
| reference_prefix	| Yes |	Path	| Absolute path to the reference genome fasta prefix. The reference genome is used by Delly for structural variant calling. |
| exclusion_file |	Yes	| Path |	Absolute path to the delly reference genome exclusion file utilized to remove suggested regions for structural variant calling. On Slurm/SGE , an HG38 exclusion file is located at /[hot\|data]/ref/hg38/delly/human.hg38.excl.tsv
| mappability_map	| Yes	| Path	| Absolute path to the delly mappability map t support GC and mappability fragment correction in CNV calling. |
| map_qual |	No |	Path |	Minimum paired-end (PE) mapping quality threshold for Delly. |
| run_qc |	No |	Boolean |	Optional parameter to indicate whether subsequent quality checks should be run. Default value is false. |
| save_intermediate_files |	Yes	| Boolean |	Optional parameter to indicate whether intermediate files will be saved. Default value is true. |
| output_dir |	Yes |	Path |	Absolute path to the directory where the output files to be saved. |
| temp_dir	| Yes	| Path |	Absolute path to the directory where the nextflow’s intermediate files are saved. |
| force_merge_duplicated_samples |	No |	Boolean	|

## Outputs

| Output |	Output type |	Description |
| ---- | ----- | -------- |
| .bcf |	Final	| Binary VCF output format with structural variants if found. |
| .bcf.csi	| Final	| CSI-format index for BCF files |
| report.html, timeline.html and trace.txt	| Log |	A Nextflow report, timeline and trace files. |
| \*.log.command.*	| Log |	Process and sample specific logging files created by nextflow. |
| *.sha512 |	Checksum |	Generates SHA-512 hash to validate file integrity. |


## Testing and Validation

### Test Data Set

Testing was performed leveraging aligned and sorted bams generated using bwa-mem2-2.1 against reference GRCh38 (SMC-HET was aligned against hs37d5):

* A-mini: BWA-MEM2-2.1_TEST0000000_TWGSAMIN000001-T001-S01-F.bam and bai
* A-partial: BWA-MEM2-2.1_TEST0000000_TWGSAPRT000001-T001-S01-F.bam and bai
* A-full: a-full-CPCG0196-B1.bam and bai
* SMC-HET: HG002.N.bam and bai

### Test runs for the A-mini/partial/full samples were performed using the following reference files

* reference_fasta: /hot/ref/hg38/genome/genome.fa
* reference_fasta_index: /hot/ref/hg38/genome/genome.fa.fai
* exclusion_file: /hot/ref/hg38/delly/human.hg38.excl.tsv
* mappability_map:/hot/ref/hg38/delly/Homo_sapiens.GRCh38.dna.primary_assembly.fa.r101.s501.blacklist.gz

## Performance Validation

Testing was performed primarily in the Boutros Lab SLURM Development cluster and the SLURM Covid cluster. Metrics below will be updated where relevant with additional testing and tuning outputs.

|Test Case	| Test Date	| Node Type |	Duration	| CPU Hours	| Virtual Memory Usage (RAM)-peak rss|
|----- | -------| --------| ----------| ---------| --------|
|A-mini	| 2021-05-12 |	F2 |	27m 47s	| 0.5h	| 1.8 GB |

## Validation Tool

Included is a template for validating your input files. For more information on the tool checkout:
https://github.com/uclahs-cds/tool-validate-nf